## STAT 207 Lab 9: Odds ratios and logistic regression

Due: Wednesday, April 8, 23:59:59

### Michael Cao AND minhcao2 

This lab consists of three main parts. It builds on material covered in 11_logistic_regression.

### Part 1: Association in a 2 x 2 table  (8 pts)

The table below summarizes data from a case-control study of the relation between smoking and heart attack. Women admitted to 30 coronary care units in Italy with acute heart attacks during 1983-1988 were each matched with 2 control patients admitted to the same hospitals with other non-heart related acute discorders. All subjects were classfiied as to whether or not they had ever smoked. The frequency counts in the different categories of the cross-classification are shown in the table. (Source: reference in A. Agresti (1996) *An Introduction to Categorical Data Analysis*, Wiley, p.26.)

In [9]:
import pandas as pd

In [10]:
# load data from knee surgury study into 2x2 table
tabl = pd.DataFrame({'HeartAttack':[172, 90], 'Control':[90, 346]}, index=['EverSmoked', 'NeverSmoked'])
tabl

HeartAttack  Control
EverSmoked           172       90
NeverSmoked           90      346

**a) (1 pts)** Calculate and display the odds of 'EverSmoked' versus 'NeverSmoked' for women in the 'HeartAttack' group.

In [11]:
p = 172 / (172 + 90)
p
odd = p/ (1-p)
odd

1.911111111111111

**b) (1 pt)** Calculate and display the odds of 'EverSmoked' versus 'NeverSmoked' in the 'Control' group.

In [12]:
p1 = 90 / (346 + 90)
odd1 = p1 / (1-p1)
odd1

0.2601156069364162

**c) (1 pt)** Calculate and display the odds ratio to compare the odds of 'EverSmoked' versus 'NeverSmoked' for the 'HeartAttack' group versus the odds of 'EverSmoked' versus 'NeverSmoked' for the 'Control' group.

In [13]:
ratio = odd / odd1
ratio

7.347160493827159

**d) (2 pts)** Calculate the log odds ratio and its standard error. Display and label them.

In [14]:
import math
log_odd_ratio = math.log(ratio)
log_odd_ratio


1.994313911210647

In [15]:
SE = math.sqrt(1/172 + 1/90 + 1/90 + 1/346)
SE

0.17585888979803932

**e) (2 pts)** Compute a 95% confidence interval for the population log-odds ratio.

In [16]:
upper_bound = (log_odd_ratio + 1.96*SE)
lower_bound = (log_odd_ratio - 1.96*SE)

In [17]:
lower_bound, upper_bound

(1.64963048720649, 2.338997335214804)

**f) (1 pt)** Interpret the result in e): does the interval include zero? Is there a statistically significant association, at level $\alpha=0.05$, between smoking status and heart attack as measured by the log odds ratio?

The interval does not include zero, there is a significant association at level $\alpha=0.05$, between smoking status and heart attack as measured by the log odds ratio

### Part 2: Lead exposure versus sex in the lead study. (8 pts)

**a) (3 pts)** The lab folder includes the csv file 'leadstudy.csv'of data from the lead exposure study. Read the data into a pandas data frame. Create a cross-classification table for the variables 'sex' and 'group'. Save and display the resulting crosstab table. 

In [18]:
import pandas as pd
df = pd.read_csv("leadstudy.csv")
df2 = df[["group", "sex"]]
tab = pd.crosstab(index = df["group"], columns 
                 = df["sex"], margins = True)
tab

sex          F   M  All
group                  
lead < 40   32  46   78
lead >= 40  16  30   46
All         48  76  124

**b) (3 pts)** Calculate and display the log-odds-ratio for high exposure versus low exposure for Male versus Female for the 2 x 2 table.

In [19]:
p_high_low_m = 30/76
p_high_low_f = 16/48
odd_m = p_high_low_m / (1 - p_high_low_m)
odd_f = p_high_low_f / (1 - p_high_low_f)
odd_m, odd_f

(0.6521739130434783, 0.49999999999999994)

In [20]:
odd_ratio = odd_m / odd_f
odd_ratio
log_odd_r = math.log(odd_ratio)
log_odd_r

0.2657031657330059

**c) (2 pts)** Calculate and display the standard error for the log-odds ratio in b).

In [23]:
SE = math.sqrt(1/32 + 1/16 + 1/46 + 1/30)
SE

0.38577514664389145

### Part 3: Logit regression of lead exposure on sex (9 pts)

**a) (3 pts)** Working with your original data frame containing the 'leadstudy.csv' data (not the crosstab table), create a new column of the data frame with a binary coding of the group exposure leve. This is a variable 'y' such that  y =1 if group == 'lead >=40' and y=0 if group == 'lead < 40'. Display the first few rows of the modiifed data frame including the new variable.

In [26]:
df2["y"] = df["group"].map({'lead < 40': 0, 'lead >= 40' : 1})
df2

/Users/minhcao/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


group sex  y
0     lead < 40   M  0
1     lead < 40   M  0
2     lead < 40   M  0
3     lead < 40   M  0
4     lead < 40   M  0
..          ...  .. ..
119  lead >= 40   F  1
120  lead >= 40   M  1
121  lead >= 40   F  1
122  lead >= 40   M  1
123  lead >= 40   M  1

[124 rows x 3 columns]

**b) (3 pts)** Use the statsmodels.formula.api function logit to fit the logistic regression of exposure (coded as the 0/1 variable 'y') with 'sex' as the explanatory variable.  Display the model summary. 

In [27]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
mod = smf.logit(formula= "y ~ sex", data = df2).fit()
mod.summary()

Optimization terminated successfully.
         Current function value: 0.657540
         Iterations 4


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                  124
Model:                          Logit   Df Residuals:                      122
Method:                           MLE   Df Model:                            1
Date:                Sat, 29 May 2021   Pseudo R-squ.:                0.002924
Time:                        16:27:06   Log-Likelihood:                -81.535
converged:                       True   LL-Null:                       -81.774
Covariance Type:            nonrobust   LLR p-value:                    0.4892
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.6931      0.306     -2.264      0.024      -1.293      -0.093
sex[T.M]       0.2657      0.386      0.689      0.491      -0.490       1.022
==============================================================================
"""

**c) (2 pt)** Compare your results in b) and c) with the 'sex' coefficient estimate and standard error in the summary table in part d). Are they giving the same or related results? Explain briefly.

They give the same result because odds ratios and logistic regression works the same when there are only 2 variable.

**d) (1 pt)** Based on your analysis of the data, is there a significant association, at level $\alpha=0.10$, between the lead exposure level (high versus low) and the sex of the subject in the study? Explain briefly.

Base on the data, we can say that there isn't any significant association at level $\alpha=0.10$, because the p value is very high : 0.49 > 0.10